File: tenant-incident/scenarios/poison-message-a/producer/producer.py
#!/usr/bin/env python3
"""producer — the normal workload.

Enqueues valid jobs onto the tail of the `jobs` list at a steady, modest rate
well under what one worker drains. This is the standing traffic, not a surge: the
backlog is healthy (near zero) until a poison message lands at the head. Holding
the producer rate constant is what makes the stall attributable to the head
message and not to load.

Env:
  REDIS_HOST  redis host (default redis)
  QUEUE       work list key (default jobs)
  RATE        jobs per second (default 2)
"""

import json
import os
import time

import redis

REDIS_HOST = os.environ.get("REDIS_HOST", "redis")
QUEUE = os.environ.get("QUEUE", "jobs")
RATE = float(os.environ.get("RATE", "2"))

r = redis.Redis(host=REDIS_HOST, port=6379, decode_responses=True)


def main():
    for _ in range(60):
        try:
            r.ping()
            break
        except Exception:
            time.sleep(1)
    interval = 1.0 / RATE if RATE > 0 else 0.5
    print(json.dumps({"event": "producing", "rate": RATE, "queue": QUEUE}), flush=True)
    i = 0
    while True:
        i += 1
        job = {"id": i, "type": "email", "payload": {"to": "user%d@example.com" % i}}
        r.rpush(QUEUE, json.dumps(job))
        if i % 20 == 0:
            print(json.dumps({"event": "produced", "count": i, "depth": r.llen(QUEUE)}), flush=True)
        time.sleep(interval)


if __name__ == "__main__":
    main()

File: tenant-incident/scenarios/poison-message-b/producer/producer.js
'use strict';

// producer — the normal workload.
//
// Inserts valid pending rows at a steady, modest rate well under what one worker
// drains, using ids starting at 1001 so a poison row injected at id 1 is always
// the head. This is the standing traffic, not a surge: the backlog is healthy
// (near zero) until a poison row lands at the head. Holding the rate constant is
// what makes the stall attributable to the head row and not to load.

const { Pool } = require('pg');

const RATE = parseFloat(process.env.RATE || '2');

const pool = new Pool({
  host: process.env.PGHOST || 'postgres',
  port: parseInt(process.env.PGPORT || '5432', 10),
  user: process.env.PGUSER || 'app',
  password: process.env.PGPASSWORD || 'app',
  database: process.env.PGDATABASE || 'jobs',
  max: 2,
});

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

async function waitForPg() {
  for (let i = 0; i < 60; i++) {
    try { await pool.query('SELECT 1'); return; } catch (e) { await sleep(1000); }
  }
}

async function main() {
  await waitForPg();
  const { rows } = await pool.query('SELECT COALESCE(MAX(id), 1000) AS m FROM jobs');
  let nextId = Math.max(1001, parseInt(rows[0].m, 10) + 1);
  const interval = RATE > 0 ? 1000 / RATE : 500;
  console.log(JSON.stringify({ event: 'producing', rate: RATE, start_id: nextId }));
  let n = 0;
  while (true) {
    const payload = JSON.stringify({ to: 'user' + nextId + '@example.com' });
    await pool.query(
      "INSERT INTO jobs (id, kind, payload, status) VALUES ($1, 'email', $2::jsonb, 'pending') ON CONFLICT (id) DO NOTHING",
      [nextId, payload]
    );
    nextId += 1;
    n += 1;
    if (n % 20 === 0) {
      const { rows: c } = await pool.query("SELECT count(*) AS n FROM jobs WHERE status = 'pending'");
      console.log(JSON.stringify({ event: 'produced', count: n, pending: parseInt(c[0].n, 10) }));
    }
    await sleep(interval);
  }
}

main();
